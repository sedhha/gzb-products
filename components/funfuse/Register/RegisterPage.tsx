import React, { useState } from 'react';
import classes from './RegisterPage.module.scss';
import WhiteLogo from '@/components/funfuse/Logos/WhiteLogo';
import WhiteButton from '@/components/funfuse/Buttons/WhiteButton/WhiteButton';
import InputBox from '@/components/common/Inputs/InputBox';
import { FcManager, FcBusinesswoman, FcBusinessman } from 'react-icons/fc';
import SelectableCard from '@/components/funfuse/SelectableCard/SelectableCard';
import { IFunFuseRegisterUser } from '@constants/interfaces/funfuse';
import { IResponse } from '@constants/interfaces/gcorn/backend/apis/response.interfaces';
import FunFuseSpinner from '@/components/common/Spinner/FunFuseSpinner';
import Head from 'next/head';

const genders = {
  male: 'Male',
  female: 'Female',
  others: 'Other',
};

type genderCardChange = (id: string) => void;
const genderCards = [
  {
    id: genders.male,
    icon: (callback: genderCardChange) => (
      <FcManager className='text-7xl' onClick={() => callback(genders.male)} />
    ),
    description: 'Man',
  },
  {
    id: genders.female,
    icon: (callback: genderCardChange) => (
      <FcBusinesswoman
        className='text-7xl'
        onClick={() => callback(genders.female)}
      />
    ),
    description: 'Woman',
  },
  {
    id: genders.others,
    icon: (callback: genderCardChange) => (
      <FcBusinessman
        className='text-7xl'
        onClick={() => callback(genders.others)}
      />
    ),
    description: 'Other',
  },
];

export default function RegisterPage() {
  const [gender, setGender] = useState(genders.male);
  const [datePickerType, setDatePickerType] = useState<string>('text');
  const [email, setEmail] = useState<string>('');
  const [password, setPassword] = useState<string>('');
  const [reEnterPassword, setReEnterPassword] = useState<string>('');
  const [userName, setUserName] = useState<string>('');
  const [name, setName] = useState<string>('');
  const [dateOfBirth, setDateOfBirth] = useState<string>(
    new Date('12-01-2001').toLocaleDateString()
  );
  const [error, setError] = useState<string>('');
  const [success, setSuccess] = useState<string>('');
  const [loading, setLoading] = useState<boolean>(false);

  const errorHandler = (error: any, message?: string) => {
    console.log(error);
    setLoading(false);
    setSuccess('');
    message
      ? setError(message)
      : setError('Client Side Unexpected Error Happened');
  };

  const successHandler = () => {
    setError('');
    setLoading(false);
    setSuccess('You have been successfully registered with FunFuse!');
  };

  const userRegistrationHandler = () => {
    setLoading(true);
    if (password !== reEnterPassword) {
      errorHandler('', 'Passwords do not match');
      return;
    }
    const registrationPayload: IFunFuseRegisterUser = {
      name,
      email,
      password,
      gender,
      dob: dateOfBirth,
      username: userName,
    };
    fetch('/api/funfuse/auth/register-user', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(registrationPayload),
    })
      .then((response) =>
        response
          .json()
          .then((data: IResponse) => {
            if (data.error) {
              errorHandler(data.opsDetails, data.opsDetails.details);
              return;
            } else {
              console.log('here');
              successHandler();
            }
          })
          .catch(errorHandler)
      )
      .catch(errorHandler);
  };

  return (
    <React.Fragment>
      <Head>
        <title>Register with Funfuse</title>
      </Head>
      <div className={classes.Card}>
        <div className={classes.LogoHeight}>
          <WhiteLogo />
        </div>
        <InputBox
          placeholder={'Name'}
          type={'text'}
          outlineColor={'#fff'}
          className='mx-4 mb-1 text-white'
          value={name}
          onChange={(e) => setName(e.target.value)}
        />
        <InputBox
          placeholder={'Email Address'}
          type={'email'}
          outlineColor={'#fff'}
          className='mx-4 mb-1 text-white'
          value={email}
          onChange={(e) => setEmail(e.target.value)}
        />
        <InputBox
          placeholder={'Unique Username'}
          type={'email'}
          outlineColor={'#fff'}
          className='mx-4 mb-1 text-white'
          value={userName}
          onChange={(e) => setUserName(e.target.value)}
        />
        <InputBox
          placeholder={'Password'}
          type={'password'}
          outlineColor={'#fff'}
          className='mx-4 mb-1 text-white'
          value={password}
          onChange={(e) => setPassword(e.target.value)}
        />
        <InputBox
          placeholder={'Re-Enter Password'}
          type={'password'}
          outlineColor={'#fff'}
          className='mx-4 mb-1 text-white'
          value={reEnterPassword}
          onChange={(e) => setReEnterPassword(e.target.value)}
        />
        <InputBox
          placeholder={'Birth Date'}
          type={datePickerType}
          onFocus={() => setDatePickerType('date')}
          onBlur={() => setDatePickerType('text')}
          outlineColor={'#fff'}
          className='mx-4 mb-1 text-white'
          style={{ colorScheme: 'dark' }}
          value={dateOfBirth}
          onChange={(e) => setDateOfBirth(e.target.value)}
        />
        <div className='flex flex-row justify-center gap-2 mt-2 width-full'>
          {genderCards.map((card) => (
            <SelectableCard
              isSelected={card.id === gender}
              key={card.id}
              icon={card.icon(setGender)}
              description={card.description}
            />
          ))}
        </div>
        {error !== '' && (
          <label className='text-center text-red-600'>{error}</label>
        )}
        {success !== '' && (
          <label className='text-center text-lime-600'>{success}</label>
        )}
        {!loading ? (
          <div className={'mt-4 text-center'}>
            <WhiteButton
              icon={
                <div
                  className={classes.LoginButton}
                  onClick={userRegistrationHandler}>
                  Register
                </div>
              }></WhiteButton>
          </div>
        ) : (
          <FunFuseSpinner />
        )}

        <br />
      </div>
    </React.Fragment>
  );
}
