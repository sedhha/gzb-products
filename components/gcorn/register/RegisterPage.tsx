import React from 'react';
import Head from 'next/head';
import Header from '../common/Header/Header';
import { IFirebaseMetaData } from '@constants/interfaces/firebase/metadata.interfaces';

interface Props {
  metadata: IFirebaseMetaData;
}

export const getDefaultGender = (gender: string[] | undefined): string =>
  gender && gender.length > 0 ? gender[0] : 'Male';
export default function GCornHomePage(props: Props) {
  const { gender } = props.metadata;
  const [selectedGender, setSelectedGender] = React.useState(
    getDefaultGender(gender)
  );
  return (
    <React.Fragment>
      <Head>
        <title>gCorn: Register</title>
        <link rel='icon' href='/gcorn/gcorn-logo.svg' />
      </Head>
      <main className='flex flex-col items-center justify-center w-screen h-screen gap-1 p-0 m-0 flex-column'>
        <Header />
        <br />

        <form className='flex flex-col items-center justify-center w-full'>
          <input
            type='email'
            placeholder='Email Address'
            className='p-2 mb-2 text-base bg-transparent border-b-2 outline-none placeholder-slate-600 border-gcorn_pl focus:border-gcorn_p'
          />
          <input
            type='password'
            placeholder='Password'
            className='p-2 mb-2 text-base bg-transparent border-b-2 outline-none placeholder-slate-600 border-gcorn_pl focus:border-gcorn_p'
          />
          <input
            type='password'
            placeholder='Confirm Password'
            className='p-2 mb-2 text-base bg-transparent border-b-2 outline-none placeholder-slate-600 border-gcorn_pl focus:border-gcorn_p'
          />
          <select
            value={selectedGender}
            onChange={(e) => setSelectedGender(e.target.value)}
            className='p-2 mb-2 text-base bg-transparent border-b-2 outline-none w-max placeholder-slate-600 border-gcorn_pl focus:border-gcorn_p'>
            {gender.map((element) => (
              <option key={element} value={element}>
                {element}
              </option>
            ))}
          </select>
          <br />
          <button
            type='submit'
            className='px-4 py-2 font-semibold bg-transparent border rounded text-gcorn_p border-gcorn_p hover:bg-gcorn_p hover:text-white hover:border-transparent'>
            Next
          </button>
        </form>
      </main>
    </React.Fragment>
  );
}
