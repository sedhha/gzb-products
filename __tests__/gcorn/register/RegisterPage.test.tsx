import RegisterPage from '@/components/gcorn/register/RegisterPage';
import { render } from '@testing-library/react';
describe('RegisterPage Test', () => {
  it('should render without throwing an error', () => {
    const { container } = render(<RegisterPage />);
    expect(container).toMatchSnapshot();
  });
});
