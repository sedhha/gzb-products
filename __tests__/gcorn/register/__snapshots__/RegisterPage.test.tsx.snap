// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterPage Test should render without throwing an error 1`] = `
<div>
  <main
    class="flex flex-col items-center justify-center w-screen h-screen gap-1 p-0 m-0 flex-column"
  >
    <div
      class="image-container"
    >
      <span
        style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
      >
        <img
          alt="gcorn-logo"
          data-nimg="fill"
          decoding="async"
          sizes="100vw"
          src="/gcorn/gcorn-logo.svg"
          srcset="/gcorn/gcorn-logo.svg 640w, /gcorn/gcorn-logo.svg 750w, /gcorn/gcorn-logo.svg 828w, /gcorn/gcorn-logo.svg 1080w, /gcorn/gcorn-logo.svg 1200w, /gcorn/gcorn-logo.svg 1920w, /gcorn/gcorn-logo.svg 2048w, /gcorn/gcorn-logo.svg 3840w"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: contain;"
        />
      </span>
    </div>
    <div
      class="flex flex-col items-center justify-center"
    >
      <h1
        class="pb-1 m-0 text-6xl italic text-gcorn_p font-gcorn_f"
      >
        gCorn!
      </h1>
      <p
        class="p-0 m-0 text-2xl italic text-gcorn_p font-gcorn_f"
      >
        A Corner for Geeks
      </p>
    </div>
    <br />
    <form
      class="flex flex-col items-center justify-center w-full"
    >
      <input
        class="p-2 mb-2 text-base bg-transparent border-b-2 outline-none placeholder-slate-600 border-gcorn_pl focus:border-gcorn_p"
        placeholder="Email Address"
        type="email"
      />
      <input
        class="p-2 mb-2 text-base bg-transparent border-b-2 outline-none placeholder-slate-600 border-gcorn_pl focus:border-gcorn_p"
        placeholder="Password"
        type="password"
      />
      <input
        class="p-2 text-base bg-transparent border-b-2 outline-none placeholder-slate-600 border-gcorn_pl focus:border-gcorn_p"
        placeholder="Confirm Password"
        type="password"
      />
      <br />
      <button
        class="px-4 py-2 font-semibold bg-transparent border rounded text-gcorn_p border-gcorn_p hover:bg-gcorn_p hover:text-white hover:border-transparent"
        type="submit"
      >
        Next
      </button>
    </form>
  </main>
</div>
`;
